Step editor improvements

elluminance opened the pull request:
I have over the past few days made a few improvements to the event/action step editor.
- Support for `SHOW_MODAL_CHOICE`, in a similar manner to `SHOW_CHOICE` (closes #246).
- `SET_MSG_EXPRESSION` is now properly supported.
- Fixed step editor from refreshing upon being updated if an event has branches/child events. (closes #294).
- Better support for the previews of a variety of types: `Vec2`, `Vec3`, `Offset`, `Entity`, `EnemyType`, `NumberExpression`, `StringExpression`, `BooleanExpression`, `VarName`, `Effect`, and `Animation`.
  - Note: This does not add any new widgets or anything to allow easier editing of these types. Merely the preview of the properties in the step editor.
- `VarCondition`s now use the string editing widget as opposed to the JSON editor. (applies to the entity editor as well).
- Player-visible text will now be properly rendered with the proper colors.

Editor before:
![image](https://github.com/CCDirectLink/crosscode-map-editor/assets/49847844/9f349269-9469-4d56-8c1c-5a13f5ff87c5)
![image-1](https://github.com/CCDirectLink/crosscode-map-editor/assets/49847844/9f7f634a-31cd-4706-84ce-6b7c18fc7f38)

Editor after:
![image](https://github.com/CCDirectLink/crosscode-map-editor/assets/49847844/90549004-a6e0-48c3-9925-ef2f1947a3a9)
![image](https://github.com/CCDirectLink/crosscode-map-editor/assets/49847844/14278cec-dcba-4596-b7f1-a6eb3f5e2197)

EL20202 commented:
Added support for text color rendering as well as support for `SET_MSG_EXPRESSION`.

Vegita2 commented on webapp/src/app/components/widgets/event-widget/event-registry/set-msg-expression.ts:
```diff
@@ -0,0 +1,32 @@
+import { Person } from '../../../../models/events';
+import { AbstractEvent, EventType } from './abstract-event';
+
+interface SetMsgExpressionData extends EventType {
```
extend `DefaultEvent` instead to avoid having to implement `getAttributes()`

Vegita2 commented on webapp/src/app/components/widgets/event-widget/event-registry/show-modal-choice.ts:
```diff
@@ -0,0 +1,78 @@
+    }[];
+}
+
+export class ShowModalChoice extends AbstractEvent<ShowModalChoiceData> {
```
extend `DefaultEvent` instead to avoid having to implement `getAttributes()`

Vegita2 commented on webapp/src/app/components/widgets/event-widget/event-registry/show-modal-choice.ts:
```diff
@@ -0,0 +1,78 @@
+			out[index] = child.events.map(v => v.export());
+		});
+		
+		return JSON.parse(JSON.stringify(out));
```
use `Helper.copy` instead to preserve the type

EL20202 replied:
That was how SHOW_CHOICE did it, so I just did it the same way (same with the 3 above things, just don't want to repeat the message), and I just fixed all that.

Vegita2 reviewed (changes requested):
Looks really good. Makes it definitely easier to use the event editor.

I noticed an issue with the new `SHOW_MODAL_CHOICE`: 
`text` shows up as `[object Object]` when text is empty, should just show nothing instead 
![image](https://github.com/CCDirectLink/crosscode-map-editor/assets/9483499/82690681-ef35-4854-82a0-3946d9ee2a4f)

EL20202 commented on webapp/src/app/components/widgets/event-widget/event-registry/abstract-event.ts:
```diff
@@ -93,6 +164,38 @@ export abstract class AbstractEvent<T extends EventType> {
+			null,
+			'#ff6969',
+			'#65ff89',
+			'#ffe430',
+			'#808080',
+			'#ffe430',
```
Oops, that was an accidental copy/paste of the wrong value. Meant to do `#ff8932`.

EL20202 commented on webapp/src/app/components/widgets/event-widget/event-registry/show-modal-choice.ts:
```diff
@@ -0,0 +1,78 @@
+    update() {
+        this.children = [];
+		this.info = this.combineStrings(
+			this.getTypeString('#7ea3ff'),
```
Is this color good? I feel like it should be a similar color but I can understand making it different.

Also I found out the source of the `[object Object]`, it was because of how `getPropString()` does not properly handle a falsy `value` being passed in. It was easy to fix though.
![image](https://github.com/CCDirectLink/crosscode-map-editor/assets/49847844/3e0da7a8-9533-48d0-b3bd-3ca037fd1957)

EL20202 commented on webapp/src/app/components/widgets/event-widget/event-registry/abstract-event.ts:
```diff
@@ -93,6 +164,38 @@ export abstract class AbstractEvent<T extends EventType> {
 		return `<span style="background-color: ${this.sanitize(color)};">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span>`;
 	}
+
+	protected getProcessedText(langLabel: Label): string {
```
There is, I just didn't want to introduce an external library where it wasn't necessary (or without permission), but if you would like I absolutely use that library.

EL20202 replied:
I just tried to implement [cc-textparser](https://github.com/redstonekasi/cc-textparser/) into the editor and whenever it reached invalid commands (even just as simple as typing in `\c` before typing in the whole command) it completely locked up the editor. I suggest not using it now unless that can be resolved - and I don't think I used it incorrectly, either. I feel in its current state, it is not meant for live text editing.

While I think using that textparser would be an improvement, I would suggest saving implementing it until later - whenever those issues with it get resolved.

lexisother replied:
@redstonekasi Any ideas?

EL20202 replied:
Gimme a few minutes, I'll make a diff so you can easily see what is happening (and maybe see if I did something wrong).

EL20202 replied:
Here, I uploaded it to discord, so you can check it out: https://ptb.discord.com/channels/382339402338402315/383264360518123520/1142964077518397460

dmitmel replied:
@EL20202 Here's a parser that handles invalid input: https://github.com/dmitmel/crosscode-translateinator/blob/c325cdad7b6f6ca40c9a73e6a98527a385a5aa4b/src/crosscode_markup.ts

redstonekasi replied:
I cannot recommend using cc-textparser right now since it was originally meant as a port of the actual CrossCode source without regard to actual usability and performance. Especially in the context of an editor such as this, not having fault tolerant parsing (is that the right term?) is not nice at all.
Not sure what is causing you issues in that example you posted on Discord since it is missing most of the relevant code.

lexisother commented:
On the topic of colours... I recently found that the ones used by the real editor are all stored in the game bundle. Why don't we copy over those? For example, this can be found in `game.feature.msg.plug-in`:
```js
wm.addStepColorRule(/SHOW_CHOICE/, "red");
```

EL20202 commented:
While I am not against that, I do feel that does go beyond the scope of my PR and should be settled separately.

Vegita2 reviewed (changes requested):
Sorry about that. I should have described it more precisely what I meant with adding types

elluminance commented on webapp/src/app/components/widgets/event-widget/event-registry/abstract-event.ts:
```diff
@@ -53,18 +54,69 @@ export abstract class AbstractEvent<T extends EventType> {
-				value = this.getVec2String(value.x, value.y);
-			} else if (type === 'Entity') {
-				value = '[' + value.name + ']';
+			switch (type as Base_WM_Type) {
```
I didn't fully understand what you meant by this, but I hope that the way I did it is what you meant. (If not, could you please provide an example?)

Vegita2 replied:
Yes, looks good. Just missing in a few cases like `'Entity'`, `EnemyType`, `VarName` etc. Basically always when a property is access from a variable typed as `any`.

Passing the `any` variable to a function that narrows it down is also good (like you did with `VarExpression`)

elluminance replied:
Alright, good - I can do that, I just wanted to do just one or two before going all in on everything and realizing "oh, that's not what was wanted." I'll probably do that within the next few days.

Vegita2 commented on webapp/src/app/components/widgets/event-widget/event-registry/show-modal-choice.ts:
```diff
@@ -0,0 +1,63 @@
+		return Helper.copy(out);
+    }
+
+    protected override generateNewDataInternal() {
+        return {
+            text: {},
+            options: [{
+				label: {
+                    en_US: 'Choice 1'
+                }
+			}, {
+				label: {
+                    en_US: 'Choice 2'
+                }
+			}]
+        };
+    }
+}
```
indentation is mixed with spaces and tabs

Vegita2 commented on webapp/src/app/services/phaser/entities/cc-entity.ts:
```diff
@@ -9,6 +9,107 @@ import { AbstractWidget } from '../../../components/widgets/abstract-widget';
+	| 'WalkAnimConfig'
+	| 'XenoDialog';
+
+//export type WMType = BaseWMType;
```
what's this?

elluminance commented:
There, I added types to what should be everything in `getPropString`. I hope that's what you wanted. I also fixed all the other issues. If there's anything else you wanted instead, let me know.
